Sources/SwiftFormatPrettyPrint/TokenStreamCreator.swift

```diff
@@ -268,6 +268,14 @@ private final class TokenStreamCreator: SyntaxVisitor {
     after(node.funcKeyword, tokens: .break)
     after(node.identifier, tokens: .close)
 
+    // prioritize keeping ") throws -> <return_type>" together.
```

Comment:
Unfortunately, this doesn't construct the token stream the way that we need it to be, and the test cases appear to be passing by coincidence (I'm actually surprised).

What's happening here is that we're at the mercy of visitation order. First, `FunctionDeclSyntax` is visited, and we call `before(node.signature.input.rightParen, tokens: .open)`. Then, after this function exits, at some pointer later we visit the children of the function decl, and that includes the `ParameterClauseSyntax` (`node.signature.input`) which also calls `before(node.rightParen, tokens: .break(.close, size: 0), .close)`.

The token stream will be created in that order, so that means you're not actually opening a new group just before the right-paren; you're opening a new group, putting a break inside it, and immediately closing it. If you run an example through the formatter with `--debug-dump-token-stream`, you can see this happening:

```
...
    [SYNTAX "(" Length: 1 Idx: 16]
    [BREAK Kind: open(kind: SwiftFormatPrettyPrint.OpenBreakKind.block) Size: 0 Length: 21 Ignores Discretionary NL: false Idx: 17]
    [OPEN Inconsistent Length: 21 Idx: 18]
      [OPEN Inconsistent Length: 8 Idx: 19]
        [SYNTAX "_" Length: 1 Idx: 20]
        [BREAK Kind: continue Size: 1 Length: 3 Ignores Discretionary NL: false Idx: 21]
        [SYNTAX "x" Length: 1 Idx: 22]
        [SYNTAX ":" Length: 1 Idx: 23]
        [BREAK Kind: continue Size: 1 Length: 4 Ignores Discretionary NL: false Idx: 24]
        [SYNTAX "Int" Length: 3 Idx: 25]
      [CLOSE Idx: 26]
---v
      [OPEN Inconsistent Length: 0 Idx: 27]
        [BREAK Kind: close(mustBreak: true) Size: 0 Length: 0 Ignores Discretionary NL: false Idx: 28]
      [CLOSE Idx: 29]
---^

      [SYNTAX ")" Length: 1 Idx: 30]
      [BREAK Kind: continue Size: 1 Length: 7 Ignores Discretionary NL: false Idx: 31]
      [SYNTAX "throws" Length: 6 Idx: 32]
      [BREAK Kind: continue Size: 1 Length: 5 Ignores Discretionary NL: false Idx: 33]
      [SYNTAX "->" Length: 2 Idx: 34]
      [SPACE Size: 1 Flexible: false Length: 1 Idx: 35]
      [SYNTAX "R" Length: 1 Idx: 36]
    [CLOSE Idx: 37]
  [CLOSE Idx: 38]
...
```

Whereas, what I think we want is this (ignore the lengths/indexes):

```
...
    [SYNTAX "(" Length: 1 Idx: 16]
    [BREAK Kind: open(kind: SwiftFormatPrettyPrint.OpenBreakKind.block) Size: 0 Length: 21 Ignores Discretionary NL: false Idx: 17]
    [OPEN Inconsistent Length: 21 Idx: 18]
      [OPEN Inconsistent Length: 8 Idx: 19]
        [SYNTAX "_" Length: 1 Idx: 20]
        [BREAK Kind: continue Size: 1 Length: 3 Ignores Discretionary NL: false Idx: 21]
        [SYNTAX "x" Length: 1 Idx: 22]
        [SYNTAX ":" Length: 1 Idx: 23]
        [BREAK Kind: continue Size: 1 Length: 4 Ignores Discretionary NL: false Idx: 24]
        [SYNTAX "Int" Length: 3 Idx: 25]
      [CLOSE Idx: 26]
---v
      [BREAK Kind: close(mustBreak: true) Size: 0 Length: 0 Ignores Discretionary NL: false Idx: 28]
    [CLOSE Idx: 29]
    [OPEN Inconsistent Length: 0 Idx: 27]
---^
      [SYNTAX ")" Length: 1 Idx: 30]
      [BREAK Kind: continue Size: 1 Length: 7 Ignores Discretionary NL: false Idx: 31]
      [SYNTAX "throws" Length: 6 Idx: 32]
      [BREAK Kind: continue Size: 1 Length: 5 Ignores Discretionary NL: false Idx: 33]
      [SYNTAX "->" Length: 2 Idx: 34]
      [SPACE Size: 1 Flexible: false Length: 1 Idx: 35]
      [SYNTAX "R" Length: 1 Idx: 36]
    [CLOSE Idx: 37]
  [CLOSE Idx: 38]
...
```

Unfortunately, this is tricky because:

* We can't do both things here because this function executes entirely before the `ParameterClauseSyntax` is visited, and we see above that that gets the ordering wrong.
* We can't do anything in the `visitPost` methods (i.e., on our way back up the tree) because by then, the token has already been claimed and added to the stream. Modifying the before/after maps at that point will be ignored.
* We could split this so that the first part happens in `visit(ParameterClauseSyntax)` and the second happens here, but we only want to do the first part if the output clause is present, and we can't determine that based on the `ParameterClauseSyntax`.

...Actually, maybe that third point isn't an issue. An `.open/.close` group surrounding just a lone `)` would be pointless, but also harmless. Could you try that (and likewise in `SubscriptDecl`)?

My other idea was going to be that we could have a way to push information down the tree by having `laterBefore/laterAfter` operations that you would start at a higher node (to keep related operations together) but then actually pick off in a lower node (to get the ordering right). That would also work and would be generally useful, but it's a bigger architectural change that we may not need right now.

Reply:
Makes sense 🤔 I confess I didn't expect the visitation order would impact when "creating" the stream, because I thought we only had to ensure we properly open/close stuff. However I now see that the ordering of the injected tokens is relevant and impacts the outcome.

> An .open/.close group surrounding just a lone ) would be pointless, but also harmless. Could you try that (and likewise in SubscriptDecl)?

Not sure if I understood correctly, but I made that change (I think) and it appears to also work, but is still not providing the final "ideal" structure. Is this what you expected? 👇 

```
    [SYNTAX "(" Length: 1 Idx: 16]
    [BREAK Kind: open(kind: SwiftFormatPrettyPrint.OpenBreakKind.block) Size: 0 Length: 21 Ignores Discretionary NL: false Idx: 17]
    [OPEN Inconsistent Length: 21 Idx: 18]
      [OPEN Inconsistent Length: 8 Idx: 19]
        [SYNTAX "_" Length: 1 Idx: 20]
        [BREAK Kind: continue Size: 1 Length: 3 Ignores Discretionary NL: false Idx: 21]
        [SYNTAX "x" Length: 1 Idx: 22]
        [SYNTAX ":" Length: 1 Idx: 23]
        [BREAK Kind: continue Size: 1 Length: 4 Ignores Discretionary NL: false Idx: 24]
        [SYNTAX "Int" Length: 3 Idx: 25]
      [CLOSE Idx: 26]
      [OPEN Inconsistent Length: 1 Idx: 27]
        [OPEN Inconsistent Length: 0 Idx: 28]
          [BREAK Kind: close(mustBreak: true) Size: 0 Length: 0 Ignores Discretionary NL: false Idx: 29]
        [CLOSE Idx: 30]
        [SYNTAX ")" Length: 1 Idx: 31]
      [CLOSE Idx: 32]
      [BREAK Kind: continue Size: 1 Length: 7 Ignores Discretionary NL: false Idx: 33]
      [SYNTAX "throws" Length: 6 Idx: 34]
      [BREAK Kind: continue Size: 1 Length: 5 Ignores Discretionary NL: false Idx: 35]
      [SYNTAX "->" Length: 2 Idx: 36]
      [SPACE Size: 1 Flexible: false Length: 1 Idx: 37]
      [SYNTAX "R" Length: 1 Idx: 38]
    [CLOSE Idx: 39]
```

Resulting from:
```swift
    // prioritize keeping ") throws -> <return_type>" together.
    if config.prioritizeKeepingFunctionOutputTogether,
      let returnTypeToken = node.signature.output?.returnType.lastToken
    {
      before(node.signature.input.rightParen, tokens: .open)
      after(node.signature.input.rightParen, tokens: .close)

      before(node.signature.input.rightParen, tokens: .open)
      after(returnTypeToken, tokens: .close)
    }
```

> My other idea was going to be that we could have a way to push information down the tree by having laterBefore/laterAfter operations that you would start at a higher node (to keep related operations together) but then actually pick off in a lower node (to get the ordering right). That would also work and would be generally useful, but it's a bigger architectural change that we may not need right now.

This sure seems like the best solution going forward. Just curious, do you mean passing additional  information on `SyntaxVisitorContinueKind.visitChildren` so that we can apply custom formatting on the next level children. or is it something different?

Reply:
I was thinking that we'd split this operation into two of the `visit` methods, so that the tokens are added in the right order with respect to when they are visited. In this one (`visit(FunctionDeclSyntax)`), we would do:

```swift
    // prioritize keeping ") throws -> <return_type>" together.
    if config.prioritizeKeepingFunctionOutputTogether {
      after(node.signature.lastToken, tokens: .close)
    }
```

Then in `visit(ParameterClauseSyntax)`, we would add the open:

```
    if config.prioritizeKeepingFunctionOutputTogether {
      before(node.rightParen, tokens: .open)
    }
```

The part about the group just around the `)` was describing what would happen in the case where the function didn't have an output clause, because we wouldn't be able to detect that down in `visit(ParameterClauseSyntax)` (without looking up through parent nodes, which I try to avoid as much as possible as a design principle, though it does happen sometimes). But then I realized that it's probably OK if that happens—groups just prioritize keeping tokens together when there are breaks between them, so a group with a single token and no breaks is basically the same as the token by itself. In other words, it should be harmless.

I haven't completely thought through whether doing this split would have any other adverse side effects, but it's the first thing that comes to mind that might get the ordering right (aside from the deeper architectural change that I mentioned).

> Just curious, do you mean passing additional information on `SyntaxVisitorContinueKind.visitChildren` so that we can apply custom formatting on the next level children. or is it something different?

We wouldn't be able to pass information down directly because we're tied to the `SyntaxVisitor` API, which says we have to return those specific enum values, and each node is visited in isolation with no direct responsibility to manage its own recursion (unless it returns `.skipChildren` and hijacks the entire walk for that node's descendants, which I want to avoid because it would make the control flow harder to reason about).

Instead, I'm thinking we could have the higher nodes in the tree associate "pending formatting tokens" with one of their descendants in a map, and then a later visit method could ask for those tokens from the map and put them in the appropriate `before/after` call. But I haven't completely thought through a good design for that.

Reply:
> The part about the group just around the ) was describing what would happen in the case where the function didn't have an output clause, because we wouldn't be able to detect that down in visit(ParameterClauseSyntax) (without looking up through parent nodes, which I try to avoid as much as possible as a design principle, though it does happen sometimes). But then I realized that it's probably OK if that happens—groups just prioritize keeping tokens together when there are breaks between them, so a group with a single token and no breaks is basically the same as the token by itself. In other words, it should be harmless.

Got it! Thanks for clearing that up! 

> I haven't completely thought through whether doing this split would have any other adverse side effects, but it's the first thing that comes to mind that might get the ordering right (aside from the deeper architectural change that I mentioned).

As mentioned on a separate comment I had to remove some checks on `before` and `after` calls (which is not ideal, but I'm unsure what's the best approach here). Besides that, it appears to be working great now. 🙌 

> We wouldn't be able to pass information down directly because we're tied to the SyntaxVisitor API, which says we have to return those specific enum values, and each node is visited in isolation with no direct responsibility to manage its own recursion (unless it returns .skipChildren and hijacks the entire walk for that node's descendants, which I want to avoid because it would make the control flow harder to reason about).

You're right, and please forgive all these silly questions from me 🤦‍♂ 

> Instead, I'm thinking we could have the higher nodes in the tree associate "pending formatting tokens" with one of their descendants in a map, and then a later visit method could ask for those tokens from the map and put them in the appropriate before/after call. But I haven't completely thought through a good design for that.

Makes sense, thanks for the explanation! 🙏 

So if I understood correctly, it would something along the lines of the already existing `pendingMultilineStringSegmentPrefixLengths = [TokenSyntax: Int]()`,
but the challenge is defining the key type for the dictionary, as we have different "families" of `Syntax` types which don't sit under a common `Hashable` parent. For the dictionary value type we could simply define an `enum` with the possible operations (position + tokens to inject), or even a closure that would contain whatever token operations we wanted (more prone to misuse). Sounds like an interesting problem!

Reply:
Can you add a comment just above the call to `after` here (and in `SubscriptDeclSyntax`), something to the effect of:

```swift
// Due to visitation order, the matching .open break is added in ParameterClauseSyntax.
```